nix/flake.nix

```
@@ -0,0 +1,148 @@
+{
+  description = "A GPU-accelerated cross-platform terminal emulator and multiplexer written by @wez and implemented in Rust";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+    rust-overlay = {
+      url = "github:oxalica/rust-overlay";
+      inputs = {
+        nixpkgs.follows = "nixpkgs";
+        flake-utils.follows = "flake-utils";
+      };
+    };
+
+    freetype2 = {
+      url = "github:wez/freetype2/de8b92dd7ec634e9e2b25ef534c54a3537555c11";
+      flake = false;
+    };
+
+    harfbuzz = {
+      url = "github:harfbuzz/harfbuzz/60841e26187576bff477c1a09ee2ffe544844abc";
+      flake = false;
+    };
+
+    libpng = {
+      url = "github:glennrp/libpng/8439534daa1d3a5705ba92e653eda9251246dd61";
+      flake = false;
+    };
+
+    zlib = {
+      url = "github:madler/zlib/cacf7f1d4e3d44d871b605da3b647f07d718623f";
+      flake = false;
+    };
+  };
+
+  outputs = {
+    self,
+    flake-utils,
+    rust-overlay,
+    nixpkgs,
+    freetype2,
+    harfbuzz,
+    libpng,
+    zlib,
+    ...
+  }:
+    flake-utils.lib.eachDefaultSystem (system: let
+      overlays = [(import rust-overlay)];
+      pkgs = import nixpkgs {inherit system overlays;};
+
+      inherit (nixpkgs) lib;
+      inherit (pkgs) stdenv;
+
+      nativeBuildInputs = with pkgs;
+        [
+          installShellFiles
+          ncurses # tic for terminfo
+          pkg-config
+          python3
+
+          rust-bin.stable.latest.minimal
+          rust-bin.stable.latest.clippy
+
+          rust-bin.nightly.latest.rustfmt
+          rust-bin.nightly.latest.rust-analyzer
+        ]
+        ++ lib.optional stdenv.isDarwin perl;
+
+      buildInputs = with pkgs;
+        [
+          fontconfig
+          pkgs.zlib
+        ]
+        ++ lib.optionals stdenv.isLinux [
+          xorg.libX11
+          xorg.libxcb
+          libxkbcommon
+          openssl
+          wayland
+          xorg.xcbutil
+          xorg.xcbutilimage
+          xorg.xcbutilkeysyms
+          xorg.xcbutilwm # contains xcb-ewmh among others
+        ]
+        ++ lib.optionals stdenv.isDarwin [
+          Cocoa
+          CoreGraphics
+          Foundation
+          libiconv
+          UserNotifications
+        ];
```

I tried building this flake on macOS - it doesn't work as it seems the `Cocoa`, `CoreGraphics` packages can't be found. 
In my own version of the flake that builds on macOS I have to provide them to `buildInputs` like so: 

```suggestion
 buildInputs = with pkgs;
        [
          fontconfig
          pkgs.zlib
        ]
        ++ lib.optionals stdenv.isLinux [
          xorg.libX11
          xorg.libxcb
          libxkbcommon
          openssl
          wayland
          xorg.xcbutil
          xorg.xcbutilimage
          xorg.xcbutilkeysyms
          xorg.xcbutilwm # contains xcb-ewmh among others
        ]
        ++ lib.optionals stdenv.isDarwin (with pkgs.darwin.apple_sdk.frameworks; [
          Cocoa
          CoreGraphics
          Foundation
          libiconv
          UserNotifications
        ]);
```

Note that I'm far from a nix expert, so I could easily be doing something wrong here.

---

I haven't tried this on MacOS, just used the code from the previous packaged wezterm derivation from nixpkgs. Is the diff you posted working without any errors or warnings for you?

---

@happenslol yeah you need the additional `with pkgs.darwin.apple_sdk.frameworks;`, since nixpkgs passes them in at the top-level nix: https://github.com/NixOS/nixpkgs/blob/486f5597f2269680a9538bfd60922489ea5dffd9/pkgs/top-level/all-packages.nix#L3128-L3130

Instead of using `with`, it's better practice to explicitly `inherit` them:
```nix
let
  inherit (pkgs.darwin.apple_sdk.frameworks) Cocoa CoreGraphics Foundation UserNotifications;
in
```
Notice that `libiconv` isn't part of `pkgs.darwin.apple_sdk.frameworks`, but `pkgs`.

---

I'm always trying to improve my nix - why is it considered better practice to inherit them instead of using `with`? A case of explicit is always better than implicit?

---

nix.dev has a best practice section about that: https://nix.dev/guides/best-practices#with-scopes
Essentially, yeah, it's about being explicit.

---

thanks!

---

Looks like it's working! I was able to build this on my macbook with 

```
nix build "github:happenslol/wezterm/add-nix-flake?dir=nix"
```

Wezterm version reported as: 
```
$ ./result/bin/wezterm --version
wezterm e941399
```